matchbox_socket/src/webrtc_socket/socket.rs

```diff
@@ -88,13 +92,90 @@ pub struct WebRtcSocketConfig {
     /// or: `wss://matchbox.example.com/your_game?next=2`
     ///
     /// The last form will pair player in the order they connect.
-    pub room_url: String,
+    pub(crate) room_url: String,
     /// Configuration for the (single) ICE server
-    pub ice_server: RtcIceServerConfig,
+    pub(crate) ice_server: RtcIceServerConfig,
     /// Configuration for one or multiple reliable or unreliable data channels
-    pub channels: Vec<ChannelConfig>,
+    pub(crate) channels: Vec<ChannelConfig>,
     /// The amount of attempts to initiate connection
-    pub attempts: Option<u16>,
+    pub(crate) attempts: Option<u16>,
+}
+
+impl WebRtcSocketBuilder {
+    /// Creates a new builder for a connection to a given room with a given number of
+    /// re-connection attempts.
+    ///
+    /// You must add at least one channel with [`WebRtcSocketBuilder::add_channel`],
+    /// [`WebRtcSocketBuilder::add_reliable_channel`], or
+    /// [`WebRtcSocketBuilder::add_unreliable_channel`] before you can build the
+    /// [`WebRtcSocket`]
+    pub fn new(room_url: impl ToString, attempts: Option<u16>) -> Self {
+        Self {
+            room_url: room_url.to_string(),
+            ice_server: RtcIceServerConfig::default(),
+            channels: Vec::default(),
+            attempts,
+        }
+    }
+
+    /// Adds a new channel to the [`WebRtcSocket`] configuration according to a [`ChannelConfig`].
+    pub fn add_channel(mut self, config: ChannelConfig) -> Self {
+        self.channels.push(config);
+        self
+    }
+
+    /// Adds a new reliable channel to the [`WebRtcSocket`].
+    ///
+    /// Messages sent via a reliable channel are guaranteed to arrive in order and will be resent
+    /// until they arrive
+    pub fn add_reliable_channel(mut self) -> Self {
+        self.channels.push(ChannelConfig::reliable());
+        self
+    }
+
+    /// Adds a new unreliable channel to the [`WebRtcSocket`].
+    ///
+    /// Messages sent via an unreliable channel may arrive in any order or not at all, but arrive as
+    /// quickly as possible
+    pub fn add_unreliable_channel(mut self) -> Self {
+        self.channels.push(ChannelConfig::unreliable());
+        self
+    }
+
+    /// Creates a [`WebRtcSocket`] and the corresponding [`MessageLoopFuture`] according to the configuration supplied.
+    ///
+    /// The returned [`MessageLoopFuture`] should be awaited in order for messages to be sent and received.
+    pub fn build(self) -> (WebRtcSocket, MessageLoopFuture) {
+        if self.channels.is_empty() {
+            panic!("You need to configure at least one channel in WebRtcSocketConfig");
```

Comment:
Insert more checks here to ensure all fields are set which don't have valid defaults, e.g. room_url, port

Reply:
Port is currently part of the room url. Would we gain anything by splitting it to protocol address/hostame and port?

Reply:
Room URL has to be parsed, making it easier to get incorrect.
For example, is `localhost:3256` correct? or is it `wss://localhost:3256`? Doesn't matter with IpAddr + u16 because it's more correct.

Reply:
Not really convinced, wouldn't we have to add *a lot* of API: ip/hostname, port, which `ws` and `wss`, path and query string building? Only to convert it to a string since that's what the browser/async_tungstenite wants anyway... I think we should just wrap `WsErr::InvalidUrl` (and the native variant) into one of our errors when connecting.

If anything, I think it would be better if the builder method for setting the url (currently the builder constructor) returned an `Err` if it was not a valid websocket url (can't be parsed by the `url` crate), or had an unexpected url scheme (not `ws` or `wss`). But I think it's debatable whether it's worth adding another dependency for this.

Regardless, kind of out of scope for this PR anyway.